bool detect_frame
bool detect_line
